README.md
[![CircleCI](https://circleci.com/gh/drewr/postal/tree/master.svg?style=svg)](https://circleci.com/gh/drewr/postal/tree/master)

postal
=======

#### Internet email library for Clojure

postal is a library for constructing and sending RFC822-compliant
Internet email messages.  It wraps the Jakarta Mail package for message
and SMTP support.  It supports sendmail natively.  Supports STARTTLS &
SSL.

### Platforms

* Anything that can run Java should work
* sendmail support likely requires *nix, but `msmtp`, et al., are worthy substitutes
* Untested on Windows

### Dependencies

* Jakarta Mail (in `lib/` after build)
* Java 8+

### Install

Served by Clojars.  In your Leiningen project.clj:

[![Clojars Project](https://img.shields.io/clojars/v/com.draines/postal.svg)](https://clojars.org/com.draines/postal)

Likewise substitute any tag name from git.

### Examples

#### Local

At a bare minimum, provide a map with `:from` and `:to` (and you'll
probably also be wanting `:subject` and `:body`, though they're
technically optional).  Any other keys you supply will show up as
ancillary headers.  This example will locally inject the message into
whatever sendmail-compatible interface your system provides.

    user> (in-ns 'postal.core)
    #<Namespace postal.core>
    postal.core> (send-message {:from "me@draines.com"
                                :to ["mom@example.com" "dad@example.com"]
                                :cc "bob@example.com"
                                :subject "Hi!"
                                :body "Test."
                                :X-Tra "Something else"})
    {:code 0, :error :SUCCESS, :message "message sent"}
    postal.core>

#### SMTP

To use SMTP, add an argument map before the message with at least
`:host` key.

    postal.core> (send-message {:host "mail.isp.net"}
                               {:from "me@draines.com"
                                :to "foo@example.com"
                                :subject "Hi!"
                                :body "Test."})
    {:code 0, :error :SUCCESS, :message "message sent"}
    postal.core>

For legacy compatibility, you can also supply these connection
parameters as metadata on the message.  `(send-message ^{:host ...} {:from ...})`

#### Authentication

Authenticate to SMTP server with `:user` and `:pass`.

    postal.core> (send-message {:host "mail.isp.net"
                                :user "jsmith"
                                :pass "sekrat!!1"}
                               {:from "me@draines.com"
                                :to "foo@example.com"
                                :subject "Hi!"
                                :body "Test."})
    {:code 0, :error :SUCCESS, :message "message sent"}
    postal.core>

#### Encryption (Gmail example)

You probably do not want to do this in the clear, so add `:ssl` to get
an encrypted connection.  This will default to port `465` if you don't
specify one.

If your destination supports TLS instead, you can use `:tls`.  This
will default to port `25`, however, so if you need a different one
make sure you supply `:port`.  (It's common for ISPs to block outgoing
port 25 to relays that aren't theirs.)

    postal.core> (send-message {:host "smtp.gmail.com"
                                :user "jsmith@gmail.com"
                                :pass "sekrat!!1"
                                :port 587
                                :tls true}
                               {:from "me@draines.com"
                                :to "foo@example.com"
                                :subject "Hi!"
                                :body "Test."})
    {:code 0, :error :SUCCESS, :message "message sent"}
    postal.core>
    
(Despite [the documentation](https://support.google.com/a/answer/176600)
mentioning SSL, your Google account is most likely restricted to TLS on port `587` and 
furthermore you may have to configure your account to
["allow less secure apps"](https://myaccount.google.com/lesssecureapps) in order to send emails.)

#### Amazon

Since Amazon SES uses authenticated SMTP, postal can use it.  Just
make sure you use a verified address and your SMTP credentials (visit
the AWS Console to set those up).  Also, if you're just sandboxing,
you can only send *to* a verified address as well.  Example:

    postal.core> (send-message {:user "AKIAIDTP........" 
                                :pass "AikCFhx1P......."
                                :host "email-smtp.us-east-1.amazonaws.com"
                                :port 587}
                   {:from "me@draines.com"  
                    :to "me@draines.com"
                    :subject "Test from Amazon SES" 
                    :body "Test!!!11"})
    {:error :SUCCESS, :code 0, :message "messages sent"}
    postal.core>

#### Attachments

Attachments and multipart messages can be added as sequences of maps:

    postal.core> (send-message {:host "mail.isp.net"}
                               {:from "me@draines.com"
                                :to "foo@example.com"
                                :subject "Hi!"
                                :body [{:type "text/html"
                                        :content "<b>Test!</b>"}
                                       ;;;; supports both dispositions:
                                       {:type :attachment
                                        :content (java.io.File. "/tmp/foo.txt")}
                                       {:type :inline
                                        :content (java.io.File. "/tmp/a.pdf")
                                        :content-type "application/pdf"}]})
    {:code 0, :error :SUCCESS, :message "message sent"}
    postal.core>

If your attachment has a content-type that is not recognized by
Jakarta Mail, e.g., `.pdf` or `.doc`, you can set `:content-type`.  You
can also set `:file-name` and `:description` if you don't like the
filename that `:content` uses.

If you want another multipart type than "mixed", you can specify it as a keyword
as the first value in the map sequence. That way you can for example create an
HTML-Email that displays a text message as fallback in email clients that do not
support (or suppress) HTML-mails:

    postal.core> (send-message {:host "localhost"
                                :port 2500
                                :user "user@localhost"
                                :pass "somePassword"}
                           {:from "jon-doe@example.com"
                            :to "jane-doe@example.com"
                            :subject "multipart/alternative test"
                            :body [:alternative
                                   {:type "text/plain"
                                    :content "This is a test."}
                                   {:type "text/html"
                                    :content "<html><head> </head><body>
                                    <h1>Heading 1</h1><p>This is a test.</p>
                                    </body></html>"}
                                  ]}))

#### UTF-8

Postal uses Jakarta Mail underneath, which defaults to charset
`us-ascii`. To set the charset, set the `:type`, like `"text/html; charset=utf-8"`.

#### Message ID

Postal will supply a message ID by default that looks like
`[random]@postal.[host]`.  You can customize this by supplying a
`:message-id` header with a function that takes no args.  The included
`postal.support/message-id` can be used if you'd like to make use of
its randomness and only customize the hostname.

    {:from "foo@bar.dom"
     :to "baz@bar.dom"
     :subject "Message IDs!"
     :body "Regards."
     :message-id #(postal.support/message-id "foo.bar.dom")}

#### User Agent

You can customize the default `User-Agent` header (by default
`postal/VERSION`).

    {:from "foo@bar.dom"
     :to "baz@bar.dom"
     :subject "Message IDs!"
     :body "Regards."
     :user-agent "MyMailer 1.0"}


#### Stress-testing

You can stress-test a server by:

    postal.core> (stress ^{:host "localhost"
                           :num     1000
                           :delay   250   ;; msecs
                           :threads 5     ;; concurrent connections}
                         {:from "foo@lolz.dom"
                          :to "bar@lolz.dom"})
    sent 1000 msgs to localhost:25
    nil
    postal.core>

### Building

    % lein deps && lein jar

## Contributors

Alexander Kouznetsov         
Allen Rohner       
Andre Branco       
Andres Cuervo     
Andy Fingerhut       
Camille Troillard       
Christoph Henkelmann       
Colin Jones       
Dante Briones      
Dimas Guardado       
Gerrit Hentschel       
J. David Lowe       
Jeff Palmucci       
Joe Gallo       
Kevin DeJong       
Kyle Kingsbury      
Paul Biggar       
Paul Stadig       
Phil Hagelberg       
Peter Onneby     
Roman Flammer       
Sam Ritchie       
Stephen Nelson         

## License

Postal is (c) 2009-2019 Andrew A. Raines and released under the MIT
license.


CHANGES
2.0.5
* Add support for supplying attachments as byte arrays

2.0.4
* Jakarta Mail 1.6.5

2.0.3
* Make make-address better for constructing addresses (#77)
* Allow bcc-only emails (#91)
* MIME-encode filenames in attachements (#95)

2.0.2

* Clean up reflection warnings (#79)

2.0.1

* Use :from for From header when both :from and :sender keys are present #47
* Avoid setting 'sender:' header when :sender key included in message #46

2.0.0

* Update branching scheme
* Add debug support on SMTP session
* Support proper JavaMail SMTPS opts

1.11.4

* Support proper URLs for attachments so we're not limited to Files
* Add generic support javamail setting strings

1.11.3

* Fix infinite loop in sendmail-send

1.11.2

* Add sendmail parameter to sendmail-send
* commons-codec 1.9
* Encourage two-arg send-message

1.11.1

* Apply charset to subject

1.11.0

* Add charset support.

1.10.3

* Fix bug where nil user/pass will trigger authentication.

1.10.2

* Eliminate many instances of reflection.

1.10.1

* Don't include clojure in dependencies.

1.10.0

* Allow custom filename and description for attachments.

1.9.0

* Support custom message ID.
* Supply default, customizable user agent.

1.8.0

* Add nested multipart support.
* Allow explicit content-type.

1.7.2

* Remove dependency on org.clojure.contrib/repl-utils 1.3.0-alpha4.





src/postal/smtp.clj
;; Copyright (c) Andrew A. Raines
;;
;; Permission is hereby granted, free of charge, to any person
;; obtaining a copy of this software and associated documentation
;; files (the "Software"), to deal in the Software without
;; restriction, including without limitation the rights to use,
;; copy, modify, merge, publish, distribute, sublicense, and/or sell
;; copies of the Software, and to permit persons to whom the
;; Software is furnished to do so, subject to the following
;; conditions:
;;
;; The above copyright notice and this permission notice shall be
;; included in all copies or substantial portions of the Software.
;;
;; THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND,
;; EXPRESS OR IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES
;; OF MERCHANTABILITY, FITNESS FOR A PARTICULAR PURPOSE AND
;; NONINFRINGEMENT. IN NO EVENT SHALL THE AUTHORS OR COPYRIGHT
;; HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER LIABILITY,
;; WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING
;; FROM, OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR
;; OTHER DEALINGS IN THE SOFTWARE.

(ns postal.smtp
  (:use [postal.message :only [make-jmessage]]
        [postal.support :only [make-props]])
  (:import [javax.mail Transport Session]))

(defn- send-single [transport ^javax.mail.Message jmsg]
  (.sendMessage transport jmsg (.getAllRecipients jmsg))
  {:code (.getLastReturnCode transport)
   :error :SUCCESS
   :message (.getLastServerResponse transport)})

(defn- send-multiple [transport jmsgs]
  (doseq [^javax.mail.Message jmsg jmsgs]
    (send-single transport jmsg))
  {:code 0 :error :SUCCESS :message "messages sent"})

(defn ^:dynamic smtp-send* [^Session session ^String proto
                            {:keys [host port user pass]} msgs]
  (assert (or (and (nil? user) (nil? pass)) (and user pass)))
  (with-open [transport (.getTransport session proto)]
    (.connect transport host port user pass)
    (let [jmsgs (map #(make-jmessage % session) msgs)]
      (if (= (count msgs) 1)
        (send-single transport (first jmsgs))
        (send-multiple transport jmsgs)))))

(defn smtp-send
  ([msg]
     (let [jmsg (make-jmessage msg)]
       (try
         (Transport/send jmsg)
         {:code 0 :error :SUCCESS :message "message sent"}
         (catch Exception e
           {:code 99 :error (class e) :message (.getMessage e)}))))
  ([args & msgs]
     (let [{:keys [host port user pass sender ssl debug]
            :or {host "localhost"}} args
            port (if (nil? port)
                   (if ssl 465 25)
                   port)
            proto (if ssl "smtps" "smtp")
            args (merge args {:port port
                              :proto proto})
            session (doto (Session/getInstance (make-props sender args))
                      (.setDebug (if debug true false)))]
       (smtp-send* session proto args msgs))))
